api/src/test/java/org/openmrs/api/TagServiceTest.java

```diff
@@ -0,0 +1,145 @@
+	
+	@Before
+	public void setupMocks() {
+		MockitoAnnotations.initMocks(this);
+	}
+	
+	@Test
+	public void getAlltags_shouldFetchAllTags() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		List<String> list = Context.getService(TagService.class).getAllTags();
+		assertEquals(13, list.size());
+	}
+	
+	@Test
+	public void getTag_shouldFetchUniqueMatchingTag() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		Tag tag = Context.getService(TagService.class).getTag(3);
+		assertEquals(tag.getUuid(), "e12c432c-1b9f-343e-b332-f3ef6c88ad3f");
+	}
+	
+	@Test
+	public void getTags_shouldFetchListOfMatchingTags() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		List<Tag> tagList = Context.getService(TagService.class).getTags("FollowUp");
+		assertEquals(tagList.size(), 4);
+	}
+	
+	@Test
+	public void object_existsMethod_shouldReturnTrueIfObjectExists() throws Exception {
+		OpenmrsObject object = Context.getService(TagService.class).getObject(Encounter.class,
+		    "e403fafb-e5e4-42d0-9d11-4f52e89d148c");
+		Encounter encounter = Context.getEncounterService().getEncounterByUuid("e403fafb-e5e4-42d0-9d11-4f52e89d148c");
+		assertEquals(object, encounter);
+	}
+	
+	@Test
+	public void getTagsWithObjectParameter_shouldReturnListOfStringTags() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		Encounter encounter = Context.getEncounterService().getEncounterByUuid("e403fafb-e5e4-42d0-9d11-4f52e89d148c");
+		List<Tag> tags = Context.getService(TagService.class).getTags(encounter);
+		assertEquals(tags.size(), 3);
+	}
+	
+	@Test
+	public void getObjectsWithAllTags_shouldReturnOpenmrsObjectsWhichHaveAllTags() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		List<String> types = new ArrayList<String>();
+		types.add("org.openmrs.Encounter");
+		types.add("org.openmrs.Obs");
+		List<String> tags = new ArrayList<String>();
+		tags.add("Initial");
+		tags.add("FollowUp");
+		List<OpenmrsObject> tagList = Context.getService(TagService.class).getObjectsWithAllTags(types, tags);
+		assertEquals(tagList.size(), 3);
+	}
+	
+	@Test
+	public void getTags_shouldReturnTagsWhichMatchParameterCriteria() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		List<String> types = new ArrayList<String>();
+		types.add("org.openmrs.Obs");
+		types.add("org.openmrs.Order");
+		List<String> tags = new ArrayList<String>();
+		tags.add("Initial");
+		tags.add("FollowUp");
+		List<Tag> tagList = Context.getService(TagService.class).getTags(types, tags);
+		assertEquals(tagList.size(), 3);
+		
+	}
+	
+	@Test
+	public void addTag_shouldAddTagToObject() throws Exception {
+		executeDataSet(TAG_INITIAL_XML);
+		Obs obs = Context.getObsService().getObsByUuid("2f616900-5e7c-4667-9a7f-dcb260abf1de");
+		Tag tag1 = Context.getService(TagService.class).addTag(obs, "Important");
+		assertEquals(tag1.getTag(), "Important");
+		assertEquals(tag1.getObjectType(), "org.openmrs.Obs");
+		assertEquals(tag1.getObjectUuid(), obs.getUuid());
```

Comment:
This test is not well written, you need to check that before you added this tag, that the call getTags(Obs) returned a list that didn't include this Obs and after it does.